### ISSUE TITLE
losing the editor's focus causes losing text selection

### ISSUE THREAD

**Issue by nebulaszlonemethi** [fraywing/textAngular#1549]
For example, when i implement a new toolbar item with custom action and i use html-based modal dialog instead of window.prompt() to ask the user for parameters, the editor will get out of focus and the text selection will be lost.

In the code, i can see some method definitions and calls of "saveFocusSelection()" and "restoreFocusSelection" put in comments saying "not functional yet" - i assume those are related.

I think solving the issue is important to be able to have real interactive actions with the user.

**Comment by luigi198** [comment]
I manage to fix this manually, I have a custom dropdown where I added all the headers as options, each time I clicked the dropdown button, I lost focus...
Look at my html

HTML
`
< div class="focus-element-editor text-angular-editor-dropdown" uib-dropdown-toggle>Format&nbsp;<i class="fa fa-caret-down"></i></div>
`

I have a custom directive containing the text-angular element, when I initialize the directive I use this init method:

JS:
`

                        // If you highlight a text and then click outside of the component, the selection is lost.
                    // With this code we 'save' the selected text, we need to add the class 'focus-element-editor'
                    // to the elements we must click to add styles from the toolbar without losing the focus.
                        var initClickEvent = function () {
                        var elementsClick = document.getElementsByClassName('focus-element-editor'), saveSelection, restoreSelection, savedSel,
                            i, n;

                        saveSelection = function() {
                            var sel = window.getSelection(), ranges = [];
                            if (sel.rangeCount) {
                                for (var x = 0, len = sel.rangeCount; x < len; ++x) {
                                    ranges.push(sel.getRangeAt(x));
                                }
                            }
                            return ranges;
                        };

                        restoreSelection = function(savedSelection) {
                            var sel = window.getSelection();
                            sel.removeAllRanges();
                            for (var x = 0, len = savedSelection.length; x < len; ++x) {
                                sel.addRange(savedSelection[x]);
                            }
                        };

                        for (i = 0, n = elementsClick.length; i<n; i++) {
                            elementsClick[i].onmousedown = function () {
                                savedSel = saveSelection();
                            };

                            elementsClick[i].onmouseup = function () {
                                restoreSelection(savedSel);
                            };
                        }
                    };`

Basically I add a class to my custom elements I want to save selection, onmousedown the element I retrieve the selection from the window object and temporarily save it, onmouseup I restore the selected text and everything works just fine, works for mobile too.

Hope it helps you

**Comment by wmai** [comment]
It works perfectly for me, thank you @luigi198
